lib/ferrum/page/tracing.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module Ferrum
+  class Page
+    class Tracing
+      INCLUDED_CATEGORIES = %w[
+        devtools.timeline
+        v8.execute
+        disabled-by-default-devtools.timeline
+        disabled-by-default-devtools.timeline.frame
+        toplevel
+        blink.console
+        blink.user_timing
+        latencyInfo
+        disabled-by-default-devtools.timeline.stack
+        disabled-by-default-v8.cpu_profiler
+        disabled-by-default-v8.cpu_profiler.hires
+      ].freeze
+      EXCLUDED_CATEGORIES = %w[
+        *
+      ].freeze
+
+      def initialize(client:)
+        self.client = client
+      end
+
+      def start(trace_params: {}, **options)
+        self.options = {
+          screenshots: false,
+          encoding: :binary,
+          **options
+        }
+        self.promise = Concurrent::Promises.resolvable_future
+        subscribe_on_tracing_event
+        inner_start(trace_params)
+      end
+
+      def stop
+        client.command("Tracing.end")
+        promise.value!
```

Comment:
Does it have timeout?

Reply:
ambiguous question because the answer is: yes and no at the same time :smile:

no - if the event is not triggered we gonna wait forever in this particular implementation.

yes - we are able to add one more param and pass it this way: `promise.value!(options[:timeout])` if we wanna stop the process when no event is triggered.
Source: https://github.com/ruby-concurrency/concurrent-ruby/blob/5b37a206e88a71939c76cd74854ab223edd56ab8/lib/concurrent-ruby/concurrent/promises.rb#L981-L992

As I understand, a file with a trace may take a long time to get in storage (due to the amount of data), that's why I think it's good to wait till the process is finished completely.